### PR TITLE
Ruby: Experimental model editor support

### PR DESCRIPTION
**Pull request by hmac**
Prototype support for working with Ruby models in the model editor.

### CODE REVIEW

File: ruby/ql/src/utils/modeleditor/ApplicationModeEndpoints.ql [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/**
+ * @name Fetch endpoints for use in the model editor (application mode)
+ * @description A list of 3rd party endpoints (methods and attributes) used in the codebase. Excludes test and generated code.
+ * @kind table
+ * @id rb/utils/modeleditor/application-mode-endpoints
+ * @tags modeleditor endpoints application-mode
+ */
+
+// This query is empty as Application Mode is not yet supported for Ruby.
+from int n
```
Comment (aibaars) [inline ruby/ql/src/utils/modeleditor/ApplicationModeEndpoints.ql:10]: Do we really need this placeholder query? If it is necessary, perhaps make it select columns of the right names and types but still use `none()` as the where clause. That should make it more clear what should be implemented if we ever support Application Mode.

File: ruby/ql/src/utils/modeleditor/ModelEditor.qll [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+      any(DataFlow::ModuleNode m | m.getOwnSingletonMethod(this.getMethodName()) = this)
+            .getQualifiedName() + "!"
+  }
+
+  /**
+   * Gets the parameter types of this endpoint.
+   */
+  bindingset[this]
+  string getParameterTypes() {
+    // For now, return the names of postional parameters. We don't always have type information, so we can't return type names.
```
Comment (aibaars) [inline ruby/ql/src/utils/modeleditor/ModelEditor.qll:70]:
```suggestion
    // For now, return the names of postional and keyword parameters. We don't always have type information, so we can't return type names.
```

File: ruby/ql/src/utils/modeleditor/ModelEditor.qll [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+  /** Holds if this API is a known sink. */
+  pragma[nomagic]
+  abstract predicate isSink();
+
+  /** Holds if this API is a known neutral. */
+  pragma[nomagic]
+  predicate isNeutral() {
+    none()
+    // this instanceof FlowSummaryImpl::Public::NeutralCallable
```
Comment (aibaars) [inline ruby/ql/src/utils/modeleditor/ModelEditor.qll:103]: Should this comment be removed ?

File: ruby/ql/src/utils/modeleditor/ModelEditor.qll [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+  string getSupportedType() {
+    this.isSink() and result = "sink"
+    or
+    this.isSource() and result = "source"
+    or
+    this.hasSummary() and result = "summary"
+    or
+    this.isNeutral() and result = "neutral"
+    or
+    not this.isSupported() and result = ""
```
Comment (aibaars) [inline ruby/ql/src/utils/modeleditor/ModelEditor.qll:125]: Is it necessary to use empty string as special value?

File: ruby/ql/src/utils/modeleditor/ModelEditor.qll [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+/**
+ * A callable where there exists a MaD source model that applies to it.
+ */
+class SourceCallable extends DataFlow::CallableNode {
+  SourceCallable() {
+    exists(string type, string path, string method |
+      method = path.regexpCapture("(Method\\[[^\\]]+\\]).*", 1) and
+      Util::pathToMethod(this, type, method) and
+      sinkModel(type, path, _)
```
Comment (aibaars) [inline ruby/ql/src/utils/modeleditor/ModelEditor.qll:164]: Should this be `sourceModel`? Currently `SinkCallable` and `SourceCallable` are identical except for the `extends` clause.
```suggestion
      sourceModel(type, path, _)
```

**Review by aibaars — COMMENTED** [review]
Some minor questions, but looks good to me otherwise.

File: ruby/ql/src/utils/modeleditor/ApplicationModeEndpoints.ql (hunk 1 above)
Comment (hmac) [reply ruby/ql/src/utils/modeleditor/ApplicationModeEndpoints.ql:10]: Yes, AFAIK the model editor expects this file to be present.

File: ruby/ql/src/utils/modeleditor/ModelEditor.qll (hunk 4 above)
Comment (hmac) [reply ruby/ql/src/utils/modeleditor/ModelEditor.qll:125]: I believe so - if we don't return a value here then the rows will not show up in the query, and most endpoints will not be supported. If you're asking why we don't use a more descriptive value such as `"not-supported"` - I don't know. But we should try to stay consistent with Java/C# as much as we can I think, and this is what they do.

File: ruby/ql/src/utils/modeleditor/ModelEditor.qll (hunk 5 above)
Comment (hmac) [reply ruby/ql/src/utils/modeleditor/ModelEditor.qll:164]: Yes! Thank you

File: ruby/ql/src/utils/modeleditor/ModelEditor.qll (hunk 3 above)
Comment (aibaars) [reply ruby/ql/src/utils/modeleditor/ModelEditor.qll]: That's fine.

**Review by aibaars — APPROVED** [review]

**Pull request merged**